contracts/TrueCurrencies/TrueRewardBackedToken.sol

```diff
@@ -230,153 +229,95 @@ contract TrueRewardBackedToken is RewardTokenWithReserve {
+        return redeemWithReserve(sender, depositAmount, _toRewardToken(depositAmount, opportunity()), opportunity());
+    }
 
-            // decrease account rewardToken balance
-            _subRewardBalance(_from, rewardAmount, finOp);
-        }
-        // 6. Sender disabled, receiver enabled, value > reserve rewardToken balance
-        // Transfer Token value between accounts and mint reward token for receiver
-        else if (!fromEnabled && toEnabled) {
-            // deposit into finOp
-            _approveAllArgs(finOp, _value, _from);
-            uint256 depositedAmount = _getFinOp(finOp).deposit(_from, _value);
-
-            // increase finOp rewardToken supply
-            finOpSupply[finOp] = finOpSupply[finOp].add(depositedAmount);
-
-            // increase account rewardToken balance
-            _addRewardBalance(_to, depositedAmount, finOp);
+    function depositWithReserve(address receiver, uint256 depositAmount, uint256 rewardAmount, address finOp) internal {
+        if (rewardTokenBalance(RESERVE, finOp) >= rewardAmount) {
+            swapTokenForReward(receiver, depositAmount, rewardAmount, finOp);
+        } else {
+            mintRewardToken(receiver, depositAmount, finOp);
         }
-        return _value;
     }
 
-    /**
-     * @dev Transfer helper function for TrueRewardBackedToken
-     */
-    function _transferAllArgs(address _from, address _to, uint256 _value) internal {
-        // 1. Both sender and receiver are disabled
-        // Exchange is in TrueCurrency -> call the normal transfer function
-        if (!trueRewardEnabled(_from) && !trueRewardEnabled(_to)) {
-            // sender not enabled receiver not enabled
-            super._transferAllArgs(_from, _to, _value);
+    function depositForReceiver(address receiver, uint256 depositAmount) internal {
+        if (!trueRewardEnabled(receiver)) {
             return;
         }
-        require(balanceOf(_from) >= _value, "not enough balance");
-
-        // require account is not blacklisted and check if hook is registered
-        (address finalTo, bool hasHook) = _requireCanTransfer(_from, _to);
 
-        _value = _transferWithRewards(_from, finalTo, _value);
+        depositWithReserve(receiver, depositAmount, _toRewardToken(depositAmount, opportunity()), opportunity());
+    }
 
-        // emit transfer event for from
-        emit Transfer(_from, _to, _value);
-        if (finalTo != _to) {
-            emit Transfer(_to, finalTo, _value);
-            if (hasHook) {
-                TrueCoinReceiver(finalTo).tokenFallback(_to, _value);
-            }
-        } else {
-            if (hasHook) {
-                TrueCoinReceiver(finalTo).tokenFallback(_from, _value);
-            }
+    function _transferWithRewards(
+        address _from,
+        address _to,
+        uint256 _value
+    ) internal returns (address) {
+        if (neitherSideHasTrueRewards(_from, _to) || isOpportunity(_from) || isOpportunity(_to)) {
```

Comment:
I would like to eliminate this piece of logic if possible. I think we should check this in the higher level function (transfer/transferFrom)